contributors/design-proposals/code-generator/kubegen.md

```diff
@@ -0,0 +1,333 @@
+for groups specified with `--api-group` will be discovered.
+
+```sh
+kubegen versioned --api-version apps/v1 --api-version apps/v1beta1 --api-version extensions/v1beta1
+```
+
+#### Manually specifying the output for the versioned client packages
+
+By default versioned client related libraries are output to the following locations:
+
+- client: `pkg/client/clientset_generated/clientset`
+- informers: `pkg/client/informers_generated/externalversions`
+- listers: `pkg/client/listers_generated`
+
+These locations can be changed using the flags `--client-output`, ``--informer-output` and `--listers-output`.
+This is useful if generating a client for external consumption.
+
+**Note**: These flags also work with `kubegen internal` since that runs generators on versioned types.
+
+```sh
+kubegen versioned --client-output kubernetes --informers-output informers --listers-output listers
+```
+
+#### Running in dry-run
+
+To verify the output without writing actual files use the `--dry-run` flag
+
+```sh
+kubegen versioned --dry-run
+```
+
+**Note**: This passes the `--verify-only` flag to each of the code generators
+
+  
+#### Overriding the searched the api directories
+
+The `--apis-dir` defaults to `pkg/apis` and looks for API groups in that directory.
+
+Looking for APIs outside the default location can be configured using
+the `--apis-dir` flag.  This flag may be provided multiple times to search multiple
+directories.
+
+```sh
+kubegen versioned --apis-dir notpkg/apis --apis-dir pkg/notapis
+```
+
+## Running via Bazel
+
+### Installing via Bazel
+
+Add the following to the project *WORKSPACE* file
+
+```py
+http_archive(
+    name = "io_k8s_rules_go",
+    url = "https://github.com/kubernetes/bazelbuild/releases/download/v1.8.0/rules_go-1.8.0.tar.gz",
+    sha256 = "<number>",
+)
+```
+
+Add the following to the project root *BUILD.bazel* file
+
+```py
+load("@io_k8s_rules_go//go:def.bzl", "code-generator")
+
+kubegen(
+    name = "kubegen",
+    cmd = "versioned",
+)
+```
+
+### Running the Bazel target
+
+```sh
+bazel run //:kubegen
+```
+
+### Bazel options
+
+- *license*: same behavior as the flag
+- *license-file*: same behavior as the flag
+- *copyright*: same behavior as the flag
+- *cmd*: controls sub command - either internal or versioned
+- *client-output*: same behavior as the flag
+- *informers-output*: same behavior as the flag
+- *listers-output*: same behavior as the flag
+
+
+## Kubernetes API and directory structure conventions
+
+kubegen will run the full set of Kubernetes code generators for all APIs it finds in the project.
+kubegen looks for API groups directories defined under `pkg/apis/<group>/<version>` where *group* and *version*
+match the patterns following Kubernetes API conventions.
+
+- **group-pattern**: ^[a-z0-9\.]+$
+- **version-pattern**: ^v\\d+(alpha\\d+|beta\\d+)?$
+
+By default, kubegen will run code generators for both versioned types defined under `pkg/apis/<group>/<version>` and
```

Comment:
Does this imply that we will move `pkg/api` to `pkg/apis/core`?

One twist is that the external types are actually under `vendor/k8s.io/api/<group>/version>`.

Reply:
> One twist is that the external types are actually under vendor/k8s.io/api/<group>/version>.

also, they're not under `<group>/<version>`, they're under `<something-vaguely-group-related>/<version>`. For example, the types for `storage.k8s.io/v1` are in the path `storage/v1`

Reply:
> Does this imply that we will move pkg/api to pkg/apis/core?

https://github.com/kubernetes/kubernetes/pull/53651

Reply:
> For example, the types for storage.k8s.io/v1 are in the path storage/v1

Are there any other patterns besides (group) and (group - domain)?

Reply:
side note: kubernetes/kubernetes#53651 merged. 

> Are there any other patterns besides (group) and (group - domain)?

We have group names and package names. They are distinct without any pattern we can assume. `<package>/<version>` is identifiers each GroupVersion uniquely. I would not switch to the logical `<GroupName>/<GroupVersion>` for specifying them on the command line.